### PR TITLE
refactor(scanner)!: Use a property that holds all nested provenances

### PR DESCRIPTION
**Pull request by sschuberth**
All known provenance information is immutable, so there is no need to have a function recreate the set on each call. Use a property instead, name it `allProvenances` to avoid the need for the `@JsonIgnore` annotation, and inline some related code for simplicity while at it.

### CODE REVIEW

File: scanner/src/main/kotlin/provenance/NestedProvenance.kt [hunk 1]
```diff
@@ -41,6 +39,15 @@ data class NestedProvenance(
      */
     val subRepositories: Map<String, RepositoryProvenance>
 ) {
+    /**
+     * Return the set of all contained [KnownProvenance]s.
+     */
+    val allProvenances: Set<KnownProvenance> =
```
Comment (fviernau) [inline scanner/src/main/kotlin/provenance/NestedProvenance.kt:45]: Does this need a `@JsonIgnore` annotation?

File: scanner/src/main/kotlin/provenance/NestedProvenance.kt [hunk 2]
```diff
@@ -41,6 +39,15 @@ data class NestedProvenance(
      */
     val subRepositories: Map<String, RepositoryProvenance>
 ) {
+    /**
```
Comment (fviernau) [inline scanner/src/main/kotlin/provenance/NestedProvenance.kt:42]: commit-message: Could it mention why this change is marked as breaking?

File: scanner/src/main/kotlin/provenance/NestedProvenanceScanResult.kt [hunk 3]
```diff
@@ -40,16 +40,11 @@ data class NestedProvenanceScanResult(
...
-     * Return a set of all [KnownProvenance]s contained in [nestedProvenance].
-     */
-    fun getProvenances(): Set<KnownProvenance> = nestedProvenance.getProvenances()
-
     /**
      * Return true if [scanResults] contains at least one scan result for each of the [KnownProvenance]s contained in
      * [nestedProvenance].
      */
-    fun isComplete(): Boolean = getProvenances().all { scanResults[it]?.isNotEmpty() == true }
+    fun isComplete(): Boolean = nestedProvenance.allProvenances.all { scanResults[it]?.isNotEmpty() == true }
```
Comment (fviernau) [inline scanner/src/main/kotlin/provenance/NestedProvenanceScanResult.kt:47]: The removal of the function in line 46 seems to me strictly speaking a bit unrelated / a thing which could become a separate commit. However, would it make sense to keep this function to not introduce these Law of Demeter violations?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/provenance/NestedProvenance.kt (hunk 1 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/provenance/NestedProvenance.kt:45]: Ouch, looks like it's indeed needed. Thanks.

Comment (fviernau) [reply scanner/src/main/kotlin/provenance/NestedProvenance.kt:45]: Commit message needs an update then.

File: scanner/src/main/kotlin/provenance/NestedProvenanceScanResult.kt (hunk 3 above)
Comment (sschuberth) [reply scanner/src/main/kotlin/provenance/NestedProvenanceScanResult.kt:47]: I agree about maybe splitting this out to a separate commit, but how is the principle of least knowledge violated here?

Comment (fviernau) [reply scanner/src/main/kotlin/provenance/NestedProvenanceScanResult.kt:47]: Just notice it's just once, the occurrence in NestedProvenanceScanResultTest.kt, so let's leave it as is.

**Review by fviernau — APPROVED** [review]
